[GSoC-2025] Blog post 4: Improved Console Output for Swift Testing (#1213)
apple/swift-org-website: _posts/2025-11-14-swift-gsoc-2025-highlight-4-swift-testing-output.md

---
layout: new-layouts/post
published: true
date: 2025-11-14 10:00:00
title: "Swift GSoC 2025 highlight: Improved Console Output for Swift Testing"
author: [ktoso, kelvin, stmontgomery]
category: "Developer Tools"
---

> Another successful year of Swift participation in [Google Summer of Code](https://summerofcode.withgoogle.com) recently came to an end, and we'd like to shine some light on the projects and work accomplished!

Reviewer: ```suggestion
The Swift community participated in [Google Summer of Code](https://summerofcode.withgoogle.com) 2025, and we've recently been showcasing all of the projects and work accomplished here on the Swift blog. You can read learn more by following these convenient links:
```

> Summer of Code is an annual program, organized by Google, which provides hands-on experience for newcomers contributing

Reviewer: ```suggestion
```

> to open source projects.

Reviewer: ```suggestion
```

> In this series of four blog posts, we'll highlight each of the Summer of Code contributors and their projects.

Reviewer: ```suggestion
```

You can navigate between the posts using these convenient links:

- [Bringing Swiftly support to VS Code](2025-11-05-swift-gsoc-2025-highlight-1-vscode-swiftly.md)

> - [JNI mode for swift-java’s source jextract tool](2025-11-07-swift-gsoc-2025-highlight-2-swift-java-jextract-jni-mode.md)

Reviewer: ```suggestion
- [Extending Swift-Java Interoperability](/blog/gsoc-2025-showcase-swift-java/)
```

> - [Improve the display of documentation during code completion in SourceKit-LSP](2025-11-12-swift-gsoc-2025-highlight-3-vscode-swift-lsp-documentation.md)

Reviewer: ```suggestion
- [Improved code completion for Swift](/blog/gsoc-2025-showcase-code-completion/)
```

> - Improved Console Output for Swift Testing (this post)

Reviewer: ```suggestion
- Improved console output for Swift Testing (this post)
```

Reviewer: Lowercases console output.

* An "issues as sub-nodes" model for displaying rich failure context directly within the hierarchy.
* An "out-dented" suite summary line that cleanly concludes each suite's output.

**Serialization-Based Architecture:**

* A future-proof design that processes `ABI.EncodedEvent` objects, preparing the reporter for an out-of-process harness architecture.
* Logic to build an in-memory representation of the test plan by consuming initial test discovery events and then looking up test details by ID for subsequent events.

**Rich Failure Reporting:**

* A dedicated `FAILED TEST DETAILS` section that provides comprehensive information for each failure, including the fully qualified test name, a detailed error message, and the exact source location (`file:line`).

## Visuals: Before & After

The Advanced Console Output Recorder fundamentally transforms how developers interact with test results. By introducing a visual hierarchy and detailed contextual information, it elevates the diagnostic experience from a flat, hard-to-parse log into a structured, scannable report.

**Current Console Output:**
![Current Console Output](/assets/images/gsoc-25/testing-output-1.jpg)

> 

Reviewer: ```suggestion
```

Reviewer: Remove extra newline in markdown.
